builtin/credential/aws/path_role.go

```diff
@@ -742,29 +811,81 @@ func (b *backend) pathRoleCreateUpdate(ctx context.Context, req *logical.Request
+	InferredEntityType          string        `json:"inferred_entity_type" structs:"inferred_entity_type" mapstructure:"inferred_entity_type"`
+	InferredAWSRegion           string        `json:"inferred_aws_region" structs:"inferred_aws_region" mapstructure:"inferred_aws_region"`
+	ResolveAWSUniqueIDs         bool          `json:"resolve_aws_unique_ids" structs:"resolve_aws_unique_ids" mapstructure:"resolve_aws_unique_ids"`
+	RoleTag                     string        `json:"role_tag" structs:"role_tag" mapstructure:"role_tag"`
+	AllowInstanceMigration      bool          `json:"allow_instance_migration" structs:"allow_instance_migration" mapstructure:"allow_instance_migration"`
+	TTL                         time.Duration `json:"ttl" structs:"ttl" mapstructure:"ttl"`
+	MaxTTL                      time.Duration `json:"max_ttl" structs:"max_ttl" mapstructure:"max_ttl"`
+	Policies                    []string      `json:"policies" structs:"policies" mapstructure:"policies"`
+	DisallowReauthentication    bool          `json:"disallow_reauthentication" structs:"disallow_reauthentication" mapstructure:"disallow_reauthentication"`
+	HMACKey                     string        `json:"hmac_key" structs:"hmac_key" mapstructure:"hmac_key"`
+	Period                      time.Duration `json:"period" mapstructure:"period" structs:"period"`
+	version                     int           `json:"version" mapstructure:"version" structs:"versoin"`
+}
+
+func (r *awsRoleEntry) ToResponseData() map[string]interface{} {
+	responseData := map[string]interface{}{
+		"auth_type":                      r.AuthType,
+		"bound_ami_id":                   r.BoundAmiIDs,
+		"bound_account_id":               r.BoundAccountIDs,
+		"bound_iam_principal_arn":        r.BoundIamPrincipalARNs,
+		"bound_iam_principal_id":         r.BoundIamPrincipalIDs,
+		"bound_iam_role_arn":             r.BoundIamRoleARNs,
+		"bound_iam_instance_profile_arn": r.BoundIamInstanceProfileARNs,
+		"bound_region":                   r.BoundRegions,
+		"bound_subnet_id":                r.BoundSubnetIDs,
+		"bound_vpc_id":                   r.BoundVpcIDs,
+		"inferred_entity_type":           r.InferredEntityType,
+		"inferred_aws_region":            r.InferredAWSRegion,
+		"resolve_aws_unique_ids":         r.ResolveAWSUniqueIDs,
+		"role_tag":                       r.RoleTag,
+		"allow_instance_migration":       r.AllowInstanceMigration,
+		"ttl":                       r.TTL / time.Second,
+		"max_ttl":                   r.MaxTTL / time.Second,
+		"policies":                  r.Policies,
+		"disallow_reauthentication": r.DisallowReauthentication,
+		"period":                    r.Period / time.Second,
+	}
+
+	convertNilToEmptySlice := func(data map[string]interface{}, field string) {
+		if data[field] == nil || len(data[field].([]string)) == 0 {
+			data[field] = []string{}
+		}
+	}
+	convertNilToEmptySlice(responseData, "bound_ami_id")
```

Comment:
Is this really necessary? The JSON marshaling code ought to see that it's a nil slice and do the right thing.

Reply:
In my testing, yes (it was a surprise to me as well). https://stackoverflow.com/a/44305910 is the best explanation I've seen of the differences between a nil slice and an empty slice. The former gets marshaled to `null` while the later gets marshaled to `[]` which makes sense to me.

Reply:
`null` is in fact the right thing, although it's untyped; or `omitempty` to just not actually include it at all. But if you think it's useful to return real data, keep it as is.